### PR TITLE
feat: Cargo (Rust) support

### LINKED ISSUE #1870: Rust (Cargo) support
#### This is a:

* [ ] Bug report (non-security related)
* [x] Feature request
* [ ] I'm not sure which of those it is

#### I'm using:

* [x] The Renovate GitHub App
* [ ] Self-hosted GitHub
* [x] Self-hosted GitLab
* [ ] Self-hosted VSTS

#### Please describe the issue:
This project seems extremely cool !
Would be even cooler with support for Rust packages (crates) ;)

I am happy to help adding the support if given some guidance/mentorship.

thank you and have a great day !

### LINKED ISSUE #2780: Add Cargo semver versioning
Necessary for #1870 

We need to add lib/versioning/semver-cargo

It will be similar to how [semver-composer](https://github.com/renovatebot/renovate/tree/master/lib/versioning/semver-composer) is implemented - i.e. some syntax is same as npm's semver but some is not.

It looks *mostly* similar. One different is the use of comma for multiples, e.g. `>= 1.2, < 1.5`

We should compare the examples listed on https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html with https://semver.npmjs.com

### PR DESCRIPTION
**Pull request by nchashch**
When finished this pull request should add support for Cargo.

TODO:
- [x] Add support for cargo's flavor of semver (`versionging/semver-cargo`)
- [x] Extract names and versions in `extractPackageFile`
- [x] Handle inline tables in `extractPackageFile`
- [x] Handle standard tables in `extractPackageFile`
- [x] Implement `updateDependency` function
- [x] Complete `readme.md`
- [x] Handle `Cargo.lock` files
- [x] Add `crates.io` datasource
- [x] Get cargo manger to update an actual crate hosted on GitLab/GitHub
- [x] Test updating inline and standard table dependencies with actual crates
- [x] Get renovate to update Cargo.lock files with actual crates
- [x] Deal with multiple versions of the same dependency in a single crate when updating Cargo.lock
- [x] Deal with workspaces and path-dependencies when updating Cargo.lock file
- [x] Support running `cargo` using docker in `getArtifacts`

Optional in this PR:
- [ ] Handle prerelease github versions
- [ ] Test with more actual crates to get better correctness guarantees

Closes #1870, #2780

### CODE REVIEW

File: lib/versioning/semver-cargo/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const semver = require('../semver');
+
+function cargo2npm(input) {
+  return input.replace(',', '||');
```
Comment (rarkins) [inline lib/versioning/semver-cargo/index.js]: Does a comma in cargo versioning mean “AND” or “OR”?

Comment (nchashch) [reply lib/versioning/semver-cargo/index.js]: It isn't specified explicitly in the reference, but it most likely means "OR". It looks like the following version requirement: ```1.2.0, > 1.4.0 <= 1.6.5``` would mean `1.2.0 OR > 1.4.0 AND <= 1.6.5`, because it works the same way for npm with `||` instead of `,` and because a space between two comparators means `AND` in the reference.

Also it wouldn't make much sense to have more than 2 terms in an `AND` expression for version requirements. For example: `> 1.2.0 AND >= 1.4.0 AND < 1.9.0` would be equivalent to `>= 1.4.0 AND < 1.9.0`.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/versioning/semver-cargo/index.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+const semver = require('../semver');
+
+function cargo2npm(input) {
```
[the remark this answers is not in the document]
Comment (nchashch) [reply lib/versioning/semver-cargo/index.js]: This is fixed in [4328e7f](https://github.com/renovatebot/renovate/pull/2795/commits/4328e7f734aaab7ca8410f38cb154d23cefe4602).

**Comment by nchashch** [comment]
@rarkins I think I could rewrite the parsing code using `toml` library, but I already have a working version of `extractPackageFile` written without using it. It saves a line number for every dependency, so with current version of `exractPackageFile` I think `updateDependency` would be best implemented by just replacing the version at the recorded line number. Although it might be a good idea to parse using the `toml` library for testing/checking for correctness.

I guess if there is a particular benefit/necessity in parsing with `toml` I could still rewrite it, but otherwise I would prefer to keep the current version.

**Comment by rarkins** [comment]
@nchashch I added a `readme.md` to this PR, can you fill it in too as part of this PR?

**Comment by nchashch** [comment]
@rarkins Yes, sure.

**Comment by nchashch** [comment]
Options for implementing [crates.io](crates.io) datasource:
1. by using `cargo search`
2. by querying [crates.io](crates.io) directly
     see [cargo source](https://github.com/rust-lang/cargo/blob/master/src/crates-io/lib.rs)
     see [API reference](https://github.com/hcpl/crates.io-http-api-reference)

**Comment by rarkins** [comment]
@nchashch query crates.io directly using their API

We have quite a few other datasources implemented already that can be used as a reference

**Comment by nchashch** [comment]
Latest version can actually update dependencies in real crates, if you enable the `cargo` package manager in `config/definitions.js`.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/cargo/artifacts.js [hunk 3]
```diff
@@ -1,28 +1,83 @@
...
+  let stderr;
+  if (updatedDeps === undefined || updatedDeps.length < 1) {
     return null;
+  }
+  const existingLockFileContent = await platform.getFile(lockFileName);
+  if (!existingLockFileContent) {
+    logger.debug('No Cargo.lock found');
+    return null;
+  }
+  const mainrs = upath.join(config.localDir, 'src/main.rs');
```
[the remark this answers is not in the document]
Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: If there is no `src/main.rs` cargo doesn't recognize a directory as a crate and exits with an error.

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/manager/cargo/artifacts.js]: So it needs both the toml and main.rs? But a totally empty/ dummy main.rs is fine?

[the remark this answers is not in the document]
Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: @rarkins Yes.

File: lib/datasource/cargo.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    return result;
+  } catch (err) {
+    if (err.statusCode === 404 || err.code === 'ENOTFOUND') {
+      logger.info({ dependency: name }, `Dependency lookup failure: not found`);
+      logger.debug({
+        err,
+      });
+      return null;
+    }
+    logger.info({ err, name }, 'crates.io lookup failure: Unknown error');
```
[the remark this answers is not in the document]
Comment (nchashch) [reply lib/datasource/cargo.js]: If 429 code is received should the datasource retry?

File: lib/manager/cargo/artifacts.js [hunk 5]
```diff
@@ -1,28 +1,83 @@
...
+  await fs.ensureDir(mainrs);
+  const localPackageFileName = upath.join(config.localDir, packageFileName);
+  const localLockFileName = upath.join(config.localDir, lockFileName);
+  try {
+    await fs.outputFile(localPackageFileName, newPackageFileContent);
+    logger.debug('Updating ' + lockFileName);
+    for (let i = 0; i < updatedDeps.length; i += 1) {
+      const dep = updatedDeps[i];
+      // Unambiguous specification of a crate would be <crate_name>:<locked_version>,
```
[the remark this answers is not in the document]
Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: Now if there are multiple versions of the same dependency, `getArtifacts` falls back to updating Cargo.lock for all dependencies with `cargo update`.

File: lib/manager/cargo/artifacts.js [hunk 6]
```diff
@@ -36,26 +36,33 @@ async function getArtifacts(
...
       const startTime = process.hrtime();
-      ({ stdout, stderr } = await exec(`${cmd} ${spec}`, {
-        shell: true,
-      }));
+      try {
+        ({ stdout, stderr } = await exec(`${cmd} ${dep}`, {
+          shell: true,
+        }));
+      } catch (err) {
+        logger.debug(
```
Comment (rarkins) [inline lib/manager/cargo/artifacts.js]: Are you sure you can do this for *all* errors. I assume you should probably against certain error messages only.

File: lib/datasource/cargo.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/datasource/cargo.js]: Normally a 429 should include a Retry-After or some indication of how long to wait. Perhaps for now we just throw an error if we get a 429, and retry for 5xx with `registry-error` thrown if it doesn't succeed after a retry.

File: lib/manager/cargo/artifacts.js [hunk 7]
```diff
@@ -1,30 +1,136 @@
...
+//   }));
+//   const duration = process.hrtime(startTime);
+//   const seconds = Math.round(duration[0] + duration[1] / 1e9);
+//   logger.info(
+//     { seconds, type: 'Cargo.lock', stdout, stderr },
+//     'Updated lockfile for ' + spec
+//   );
+// }
+
+async function updateAll(localPackageFileName) {
```
Comment (rarkins) [inline lib/manager/cargo/artifacts.js]: When is this used?

**Comment by rarkins** [comment]
@nchashch could you create a new PR that is just for the lib/datasource/cargo.js part? (including its tests). My goal is to merge it first (even if it's unused for now).

I would like to review and merge it first, and then we merge it into this branch to reduce its number of modified files.

**Comment by nchashch** [comment]
@rarkins Ok. I am on it.

**Comment by rarkins** [comment]
@nchashch the cargo datasource is now merged. Can you also split off the cargo versioning into a separate PR so that it can be reviewed and merged separately?

**Comment by nchashch** [comment]
@rarkins Yes, sure.

File: lib/manager/cargo/readme.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+#### Can package files have "local" links to each other that need to be resolved?
+
+It is possible to have local dependencies, by specifying a file path.
+
+---
+
+#### Is there reason why package files need to be parsed together (in serial) instead of independently?
+
+No a single Cargo.toml file specifies a single package.
```
Comment (rarkins) [inline lib/manager/cargo/readme.md]: Wouldn't Cargo's "workspaces" concept mean we need to treat them as an interlinked group and not just as separate/autonomous packages?

Comment (nchashch) [reply lib/manager/cargo/readme.md]: > A workspace is a set of packages that share the same Cargo.lock and output directory.

see [reference](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-a-workspace)

It looks like if there is a workspace, it would only affect `getArtifacts` function. We would have to run `cargo update` inside an actual repository, or at least in a directory with proper file structure.
Instead of running `cargo update` for each `Cargo.toml` separately if the corresponding `Cargo.lock` exists.

Comment (rarkins) [reply lib/manager/cargo/readme.md]: It means we need to be able to locate the “root” Cargo.toml, which doesn’t necessarily need to be at the root level of the repository, right? But I think for launch you were planning to support only projects with a Cargo.toml at the root anyway?

Comment (nchashch) [reply lib/manager/cargo/readme.md]: > It means we need to be able to locate the “root” Cargo.toml, which doesn’t necessarily need to be at the root level of the repository, right?

Yes, most likely. But identifying "root" Cargo.toml seems to be an easy problem, since it would contain a `[workspace]` section, see [reference](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-a-workspace).

Also it looks like `cargo` treats path dependencies similarly to members of a workspace, it keeps a single `Cargo.lock` file in root directory. And in order to run `cargo update` it needs to have `Cargo.toml` files for them (located at correct paths).

Comment (rarkins) [reply lib/manager/cargo/readme.md]: So do we need to run cargo update only in the root workspace? Or run it in the current directory of whatever toml file is updated?

Comment (nchashch) [reply lib/manager/cargo/readme.md]: It doesn't matter, you can run `cargo update` from root directory or any subdirectory of the crate,
It must have proper file structure. like:
```
├── Cargo.lock
├── Cargo.toml
├── add-one
│   ├── Cargo.toml
│   └── src
│       └── lib.rs
├── adder
│   ├── Cargo.toml
│   └── src
│       └── main.rs
└── target
```
It would probably be enough to create empty `src/main.rs` files in order to make subcrates valid, like in current version.

**Comment by nchashch** [comment]
I tested current version on a crate with a workspace and path dependencies, and it does update all the `Cargo.toml` files in nested crates, but it doesn't update the `Cargo.lock` in root crate (which is supposed to be shared by all nested crates). It will definitely have to be fixed before merging.

**Comment by rarkins** [comment]
@nchashch 

> I tested current version on a crate with a workspace and path dependencies, and it does update all the `Cargo.toml` files in nested crates, but it doesn't update the `Cargo.lock` in root crate (which is supposed to be shared by all nested crates). It will definitely have to be fixed before merging.

Can you share this repo, or create something similar that I can test against?

**Comment by nchashch** [comment]
@rarkins
Yes, sure: https://gitlab.com/nchashch/workspace_example

**Comment by rarkins** [comment]
@nchashch are you sure this is a working example?

I see this error when running Cargo:

```
error: failed to parse manifest at `/var/folders/1y/nbqcg2hn73dg7bc9vpkbn18r0000gn/T/renovate/gitlab/renovate-tests/workspace_example/crate5/Cargo.toml`

Caused by:
  no targets specified in the manifest
  either src/lib.rs, src/main.rs, a [lib] section, or [[bin]] section must be present",
```

Also I was expecting to see Artifacts errors on some of the MRs but I'm not.

**Comment by nchashch** [comment]
>error: failed to parse manifest at `/var/folders/1y/nbqcg2hn73dg7bc9vpkbn18r0000gn/T/renovate/gitlab/renovate-tests/workspace_example/crate5/Cargo.toml`
Caused by:
  no targets specified in the manifest
  either src/lib.rs, src/main.rs, a [lib] section, or [[bin]] section must be present",

This error should be fixed by [cargo: Add createMainrs function and call it in getArtifacts](/renovatebot/renovate/pull/2795/commits/3b8696cb44b02c8617b865a5dbcafb2c9e901c47).

> Also I was expecting to see Artifacts errors on some of the MRs but I'm not.

You need to use the latest version of this branch. Because before [cargo: Fix lockFileName in getArtifacts](/renovatebot/renovate/pull/2795/commits/f9f8354c56acc37ac3f03880ec85d1981151972d)
 commit `getArtifacts` wasn't triggered for subcrates.

**Comment by rarkins** [comment]
@nchashch I'm pretty sure I'm running the latest commit on this branch. I did a merge from master too. Can you pull the latest from this branch and test it out locally?

**Comment by nchashch** [comment]
This project has a workspace and a lot of path dependencies: https://github.com/amethyst/amethyst
Also this project is organized into subcrates using a workspace: https://github.com/gfx-rs/gfx

**Comment by nchashch** [comment]
The reason Cargo gives this error:
> error: failed to parse manifest at `/var/folders/1y/nbqcg2hn73dg7bc9vpkbn18r0000gn/T/renovate/gitlab/renovate-tests/workspace_example/crate5/Cargo.toml`
> Caused by:
> no targets specified in the manifest
> either src/lib.rs, src/main.rs, a [lib] section, or [[bin]] section must be present",

is that not all subcrates are written to `config.localDir` when first 5 updates happen.

`cargo update` requires `Cargo.toml` files with stub `src/main.rs` files to exist for every path dependency and member of the workspace in order to run.

> Also I was expecting to see Artifacts errors on some of the MRs but I'm not.

The last two updates don't error out because all 5 subcrates were created, and `cargo update` can run properly. You can see this by changing
```rust
logger.trace('Created file', fileName);
```
to
```rust
logger.info(`Created file ${fileName}`);
```
at line 158 in `lib/manager/cargo/artifacts.js`. In first 5 updates there would be less than 5 `Created file ...` lines in the log.

**Comment by rarkins** [comment]
OK, let's switch to using gitFs only for Cargo. It's a little harder to set up on GitLab (needs SSH key too) but easy on the rest. With gitFs, the config.localDir will always have all the files from the repo without having to worry about what to copy.

In such a case that we are doing a full clone of the repo, what changes would you make to your Cargo code for Renovate? e.g. we no longer have to write fake `main.rs` files, right?

**Comment by nchashch** [comment]
>In such a case that we are doing a full clone of the repo, what changes would you make to your Cargo code for Renovate? e.g. we no longer have to write fake `main.rs` files, right?

Yes, if there is a clone of the repository in `config.localDir`, there is no need for writing fake `main.rs` files. I think it would only affect the `getArtifacts` function. Datasource, versioning, `extractPackageFile`, and `updateDependency` should work the same.

**Comment by nchashch** [comment]
There is also an unrelated bug (most likely introduced by me merging master) in current version of `getArtifacts`: it returns just an object instead of a `[ list ]` of objects, I am fixing it in the following commit.

I have tested fixed version it and it does update `Cargo.lock` properly for crates without workspaces.

File: lib/manager/cargo/artifacts.js [hunk 9]
```diff
@@ -1,27 +1,117 @@
...
+  const localPackageFileName = upath.join(config.localDir, packageFileName);
+  const localLockFileName = upath.join(config.localDir, lockFileName);
+  let stdout;
+  let stderr;
+  try {
+    await fs.outputFile(localPackageFileName, newPackageFileContent);
+    logger.debug('Updating ' + lockFileName);
+    for (let i = 0; i < updatedDeps.length; i += 1) {
+      const dep = updatedDeps[i];
+      // TODO: document this command and why we use it
```
Comment (rarkins) [inline lib/manager/cargo/artifacts.js]: @nchashch could you add some documentation of these parameters and remove this TODO?

Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: Sure.

File: lib/manager/cargo/artifacts.js [hunk 10]
```diff
@@ -1,27 +1,117 @@
...
+      //   cmd = 'docker run renovate/cargo update -p';
+      // }
+      const startTime = process.hrtime();
+      try {
+        ({ stdout, stderr } = await exec(`${cmd} ${dep}`, {
+          shell: true,
+        }));
+      } catch (err) /* istanbul ignore next */ {
+        const msgStart = 'error: There are multiple';
+        // TODO: Document why we need to do this
```
Comment (rarkins) [inline lib/manager/cargo/artifacts.js]: @nchashch could you add some documentation of why we catch/retry this?

File: lib/manager/cargo/artifacts.js (hunk 9 above)
Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: I have added the documentation.

File: lib/manager/cargo/artifacts.js (hunk 10 above)
Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: Done.

File: lib/manager/cargo/artifacts.js [hunk 11]
```diff
@@ -1,27 +1,126 @@
...
+        }));
+      } catch (err) /* istanbul ignore next */ {
+        // Two different versions of one dependency can be present in the same
+        // crate, and when that happens an attempt to update it with --package ${dep}
+        // key results in cargo exiting with error code `101` and an error mssage:
+        // "error: There are multiple `${dep}` packages in your project".
+        //
+        // If exception `err` was caused by this, we execute `updateAll` function
+        // instead of returning an error. `updateAll` function just executes
+        // "cargo update --manifest-path ${localPackageFileName}" without the `--package` key.
```
Comment (rarkins) [inline lib/manager/cargo/artifacts.js]: Does this mean that every package gets updated, not just the ones we want?

Comment (rarkins) [reply lib/manager/cargo/artifacts.js]: Also, do you have an example of this type of file?

Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: > Does this mean that every package gets updated, not just the ones we want?

Yes, it updates all packages in `Cargo.lock`, which are out of sync with `Cargo.toml`.

> Also, do you have an example of this type of file?

Yes: https://gitlab.com/nchashch/cargo_lock
If you try running
```
cargo update -p log
```
it will error out.

Comment (rarkins) [reply lib/manager/cargo/artifacts.js]: Why would it error? I only see `log` in there once

Comment (rarkins) [reply lib/manager/cargo/artifacts.js]: BTW if I clone that repo and run `cargo install` then it errors, e.g.

```
error[E0603]: module `stdio` is private
   --> /Users/rhys/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.1.10/src/lib.rs:226:36
    |
226 |     handle: LineBufferedWriter<io::stdio::StdWriter>,
    |
```

Am I doing something wrong?

Comment (rarkins) [reply lib/manager/cargo/artifacts.js]: `cargo update -p log` fails as you suggest. It's possible to get it to succeed if I run `cargo update -p log:0.1.0` (i.e. the old value).

`cargo update` works, but it also updates unrelated packages such as `libc`.

e.g.

```diff
 [[package]]
 name = "libc"
-version = "0.2.48"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-
-[[package]]
-name = "log"
-version = "0.1.10"
+version = "0.2.49"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-dependencies = [
- "regex 0.1.80 (registry+https://github.com/rust-lang/crates.io-index)",
-]

 [[package]]
 name = "log"
@@ -71,7 +63,7 @@ name = "memchr"
 version = "0.1.11"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
- "libc 0.2.48 (registry+https://github.com/rust-lang/crates.io-index)",
+ "libc 0.2.49 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
```

Running `cargo generate-lockfile` produce the same diff as `cargo update`

Comment (nchashch) [reply lib/manager/cargo/artifacts.js]: > BTW if I clone that repo and run `cargo install` then it errors, e.g.
> 
> ```
> error[E0603]: module `stdio` is private
>    --> /Users/rhys/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.1.10/src/lib.rs:226:36
>     |
> 226 |     handle: LineBufferedWriter<io::stdio::StdWriter>,
>     |
> ```
> 
> Am I doing something wrong?

That repo has very old versions of crates as deps (`log 0.1.0` was released on December 14, 2014), so it most likely fails to build because recent versions of rust can't build them anymore.

**Comment by pyrrho** [comment]
Hey all. A friend of mine and I set up a dev copy of this branch to see how it behaved in a relatively complex Rust workspace, in a bare-metal CI setup (very pleased with the results), and came across one issue that we'd like to raise.

(Let me quickly say that I'd like to spend -- I dunno -- three paragraphs gushing about this addition, and this tool in general. For the sake of expediency I won't, but please pretend like I did.)

tl;dr - [cargo yank](https://doc.rust-lang.org/cargo/commands/cargo-yank.html) isn't something renovate is currently aware of.

Slightly longer description - crates.io strongly discourages the deletion of published crate versions -- even those that include bugs or broken APIs -- but they do provide a mechanism for publishers to blacklist versions of their crates. This allows users of "yanked" versions (those that are already in Cargo.lock) to continue using the problematic version, but will cause `cargo update` to skip the yanked version and trigger an error if a yanked version is being _added_ to a Cargo.toml or a Cargo.lock.

Specific to my purposes, the dev version of renovate I'm running attempted to update [handlebars](https://crates.io/crates/handlebars) from 1.1.0 to 1.2.0, even though 1.2.0 has been yanked. This lead to a PR that couldn't be merged.

I believe the only change necessary to make renovate aware of yanked versions would be to add a filter pass [somewhere around here](https://github.com/renovatebot/renovate/blob/62fc503/lib/datasource/cargo/index.js#L20) to check if `version.yanked === true`. (I've provided an abridged copy of crate.io's listing for handlebars below, including the yanked version, for reference.)

I'm aware that this change would expand the scope of this PR, and I'd like to apologize if any/all would have preferred an issue over this (novel-length) comment. I figured that, because this is ongoing work, the PR would be a fine place to raise concerns. I'm aware that's somewhat a mater of preference, so I'm hoping all parties involved are okay with it.

Thanks for your work on this!

##### `curl https://crates.io/api/v1/crates/handlebars`
```
{
  "crate": {
    "id": "handlebars",
    . . .
    "versions": [122401, 114332, 110862, . . . ],
    . . .
    "max_version": "1.1.0",
  },
  "versions": [
    {
      "id": 122401,
      "crate": "handlebars",
      "num": "1.2.0",
      . . .
      "yanked": true,
      . . .
    },
    {
      "id": 114332,
      "crate": "handlebars",
      "num": "1.1.0",
      . . .
      "yanked": false,
      . . .
    },
    . . .
  ],
  . . .
}
```

**Comment by rarkins** [comment]
Hi @pyrrho, thanks very much for testing out this PR! I have implemented yank-awareness into this branch so that it's ready from day one. I tested it out on handlebars as you suggest and it works.

I've also added docker support now so I think we are ready to merge this to master. @nchashch do you agree?

**Comment by rarkins** [comment]
image
